src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3215,7 +3146,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           // less expensive than including them in inferMethodAlternative (see below).
           def shapeType(arg: Tree): Type = arg match {
             case Function(vparams, body) =>
-              functionType(vparams map (_ => AnyTpe), shapeType(body))
+              functionType(vparams map (_ => AnyTpe), shapeType(body)) // TODO: should this be erased when retyping during erasure?
```

Comment:
No wait, `implSym` is only used in a assertion failure message. False alarm.

Tested with:

```
scala> trait T { private def foo(a: Option[Int], b: Any) = 1; private def foo(a: Option[String], b: String) = 2; def bar = foo(Some[Int](1), "") }
defined trait T

scala> new T{}.bar
res1: Int = 1
```